File: fleet-service/tracing.py
import os
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import SERVICE_NAME, Resource
from opentelemetry.instrumentation.flask import FlaskInstrumentor
from opentelemetry.instrumentation.requests import RequestsInstrumentor


def init_jaeger_tracing(service_name):
    """Initialize Jaeger tracing for the Flask application"""
    jaeger_endpoint = os.getenv("JAEGER_ENDPOINT", "http://localhost:4318")
    
    # Create OTLP exporter
    otlp_exporter = OTLPSpanExporter(endpoint=f"{jaeger_endpoint}/v1/traces")
    
    # Create trace provider
    trace_provider = TracerProvider(
        resource=Resource.create({
            SERVICE_NAME: service_name,
        })
    )
    
    # Add span processor
    trace_provider.add_span_processor(BatchSpanProcessor(otlp_exporter))
    
    # Set global trace provider
    trace.set_tracer_provider(trace_provider)
    
    return trace_provider


def instrument_app(app, service_name):
    """Instrument a Flask app with tracing"""
    # Initialize Jaeger
    tp = init_jaeger_tracing(service_name)
    
    # Instrument Flask
    FlaskInstrumentor().instrument_app(app)
    
    # Instrument requests library
    RequestsInstrumentor().instrument()
    
    return tp

File: shelter-service/src/tracing.js
const { NodeSDK } = require("@opentelemetry/sdk-node");
const { getNodeAutoInstrumentations } = require("@opentelemetry/auto-instrumentations-node");
const { OTLPTraceExporter } = require("@opentelemetry/exporter-otlp-proto-http");
const { Resource } = require("@opentelemetry/resources");
const { SemanticResourceAttributes } = require("@opentelemetry/semantic-conventions");

function initJaegerTracing(serviceName) {
  const jaegerEndpoint = process.env.JAEGER_ENDPOINT || "http://localhost:4318";

  const sdk = new NodeSDK({
    resource: new Resource({
      [SemanticResourceAttributes.SERVICE_NAME]: serviceName,
    }),
    traceExporter: new OTLPTraceExporter({
      url: `${jaegerEndpoint}/v1/traces`,
    }),
    instrumentations: [getNodeAutoInstrumentations()],
  });

  sdk.start();
  console.log(`Jaeger tracing initialized for ${serviceName}`);

  process.on("SIGTERM", () => {
    sdk
      .shutdown()
      .then(() => console.log("Tracing terminated"))
      .catch((log) => console.log("Error terminating tracing", log))
      .finally(() => process.exit(0));
  });

  return sdk;
}

module.exports = { initJaegerTracing };
